File: relay/message_sender.js
// vim: ts=4:sw=4:expandtab


const Attachment = require('./attachment');
const OutgoingMessage = require('./outgoing_message');
const crypto = require('./crypto');
const errors = require('./errors');
const eventing = require('./eventing');
const hub = require('./hub');
const libsignal = require('libsignal');
const node_crypto = require('crypto');
const protobufs = require('./protobufs');
const queueAsync = require('./queue_async');
const storage = require('./storage');
const uuid4 = require('uuid/v4');

class Message {

    constructor(options) {
        Object.assign(this, options);
        if (typeof this.timestamp !== 'number') {
            throw new Error('Invalid timestamp');
        }
        if (this.expiration !== undefined && this.expiration !== null) {
            if (typeof this.expiration !== 'number' || !(this.expiration >= 0)) {
                throw new Error('Invalid expiration');
            }
        }
        if (this.attachments) {
            if (!(this.attachments instanceof Array)) {
                throw new Error('Invalid message attachments');
            }
        }
        if (this.flags !== undefined && typeof this.flags !== 'number') {
            throw new Error('Invalid message flags');
        }
    }

    isEndSession() {
        return (this.flags & protobufs.DataMessage.Flags.END_SESSION);
    }

    toProto() {
        const dataMessage = protobufs.DataMessage.create();
        if (this.body) {
            dataMessage.body = JSON.stringify(this.body);
        }
        if (this.attachmentPointers && this.attachmentPointers.length) {
            dataMessage.attachments = this.attachmentPointers;
        }
        if (this.flags) {
            dataMessage.flags = this.flags;
        }
        if (this.expiration) {
            dataMessage.expireTimer = this.expiration;
        }
        //return protobufs.Content.encode(protobufs.Content.create({dataMessage})).finish();
        return protobufs.Content.create({dataMessage});
    }
}


class MessageSender extends eventing.EventTarget {

    constructor({addr, signal, atlas}) {
        super();
        this.addr = addr;
        this.signal = signal;
        this.atlas = atlas;
    }

    static async factory() {
        const addr = await storage.getState('addr');
        const signal = await hub.SignalClient.factory();
        const atlas = await hub.AtlasClient.factory();
        return new this({addr, signal, atlas});
    }

    async makeAttachmentPointer(attachment) {
        if (!(attachment instanceof Attachment)) {
            throw TypeError("Expected `Attachment` type");
        }
        const key = node_crypto.randomBytes(64);
        const ptr = protobufs.AttachmentPointer.create({
            key,
            contentType: attachment.type
        });
        const iv = node_crypto.randomBytes(16);
        const encryptedBin = await crypto.encryptAttachment(attachment.buffer, key, iv);
        ptr.id = await this.signal.putAttachment(encryptedBin);
        return ptr;
    }

    async uploadAttachments(message) {
        const attachments = message.attachments;
        if (!attachments || !attachments.length) {
            message.attachmentPointers = [];
            return;
        }
        const uploads = attachments.map(x => this.makeAttachmentPointer(x));
        try {
            message.attachmentPointers = await Promise.all(uploads);
        } catch(e) {
            if (e instanceof errors.ProtocolError) {
                throw new errors.MessageError(message, e);
            } else {
                throw e;
            }
        }
    }

    async send({
        to=null, distribution=null,
        text=null, html=null, body=[],
        data={},
        threadId=uuid4(),
        threadType='conversation',
        threadTitle=undefined,
        messageType='content',
        messageId=uuid4(),
        messageRef=undefined,
        expiration=undefined,
        attachments=undefined,
        flags=undefined,
        sendTime=new Date(),
        userAgent='librelay'
    }) {
        if (!distribution) {
            if (!to) {
                throw TypeError("`to` or `distribution` required");
            }
            distribution = await this.atlas.resolveTags(to);
        }
        if (text) {
            body.push({
                type: 'text/plain',
                value: text
            });
        }
        if (html) {
            body.push({
                type: 'text/html',
                value: html
            });
        }
        if (body.length) {
            data.body = body;
        }
        if (attachments && attachments.length) {
            data.attachments = attachments.map(x => x.getMeta());
        }
        const timestamp = sendTime.getTime();
        const msg = new Message({
            addrs: distribution.userids,
            threadId,
            body: [{
                version: 1,
                threadId,
                threadType,
                messageId,
                messageType,
                messageRef,
                distribution: {
                    expression: distribution.universal
                },
                sender: {
                    userId: this.addr
                },
                sendTime: sendTime.toISOString(),
                userAgent,
                data
            }],
            timestamp,
            attachments,
            expiration,
            flags
        });
        await this.uploadAttachments(msg);
        const msgProto = msg.toProto();
        await this._sendSync(msgProto, timestamp, threadId, expiration && Date.now());
        return this._send(msgProto, timestamp, this.scrubSelf(distribution.userids));
    }

    _send(msgProto, timestamp, addrs) {
        console.assert(addrs instanceof Array);
        const outmsg = new OutgoingMessage(this.signal, timestamp, msgProto);
        outmsg.on('keychange', this.onKeyChange.bind(this));
        for (const addr of addrs) {
            queueAsync('message-send-job-' + addr, () =>
                outmsg.sendToAddr(addr).catch(this.onError.bind(this)));
        }
        return outmsg;
    }

    async onError(e) {
        const ev = new eventing.Event('error');
        ev.error = e;
        await this.dispatchEvent(ev);
    }

    async onKeyChange(e) {
        await this.dispatchEvent(new eventing.KeyChangeEvent(e));
    }

    async _sendSync(content, timestamp, threadId, expirationStartTimestamp) {
        const sentMessage = protobufs.SyncMessage.Sent.create({
            timestamp,
            message: content.dataMessage
        });
        if (threadId) {
            sentMessage.destination = threadId;
        }
        if (expirationStartTimestamp) {
            sentMessage.expirationStartTimestamp = expirationStartTimestamp;
        }
        const syncMessage = protobufs.SyncMessage.create({sent: sentMessage});
        const syncContent = protobufs.Content.create({syncMessage});
        return this._send(syncContent, timestamp, [this.addr]);
    }

    async syncReadMessages(reads) {
        if (!reads.length) {
            console.warn("No reads to sync");
        }
        const read = reads.map(r => protobufs.SyncMessage.Read.create({
            timestamp: r.timestamp,
            sender: r.sender
        }));
        const syncMessage = protobufs.SyncMessage.create({read});
        const content = protobufs.Content.create({syncMessage});
        return this._send(content, Date.now(), [this.addr]);
    }

    scrubSelf(addrs) {
        const nset = new Set(addrs);
        nset.delete(this.addr);
        return Array.from(nset);
    }

    async closeSession(addr, timestamp=Date.now()) {
        const dataMessage = protobufs.DataMessage.create({
            flags: protobufs.DataMessage.Flags.END_SESSION
        });
        const content = protobufs.Content.create({dataMessage});
        const outmsg = this._send(content, timestamp, [addr]);
        const deviceIds = await storage.getDeviceIds(addr);
        await new Promise(resolve => {
            outmsg.on('sent', resolve);
            outmsg.on('error', resolve);
        });
        await Promise.all(deviceIds.map(deviceId => {
            const address = new libsignal.SignalProtocolAddress(addr, deviceId);
            const sessionCipher = new libsignal.SessionCipher(storage, address);
            return sessionCipher.closeOpenSessionForDevice();
        }));
    }
}

module.exports = MessageSender;

File: relay/message_sender.py
"""
Forsta message sending interface.
"""

import asyncio
import datetime
import json
import logging
import secrets
import uuid
from . import crypto, eventing, hub, storage
from . import protobufs
from .attachment import Attachment
from .outgoing_message import OutgoingMessage

store = storage.getStore()
logger = logging.getLogger(__name__)


def msnow():
    """ Epoch in ms like java*. """
    return round(datetime.datetime.now().timestamp() * 1000)


class Message(object):
    """ XXX This is a silly interface right now..  prove your worth! """

    def __init__(self, **options):
        self.__dict__.update(options)  # XXX ick.. 

    def isEndSession(self):
        return self.flags & protobufs.DataMessage.END_SESSION

    def toProto(self):
        content = protobufs.Content()
        #dataMessage = protobufs.DataMessage()
        dataMessage = content.dataMessage
        if getattr(self, 'body', None):
            dataMessage.body = json.dumps(self.body)
        if getattr(self, 'attachmentPointers', None):
            dataMessage.attachments = self.attachmentPointers
        if getattr(self, 'flags', None):
            dataMessage.flags = self.flags
        if getattr(self, 'expiration', None):
            dataMessage.expireTimer = self.expiration
        #content = protobufs.Content()
        #content.dataMessage = dataMessage
        return content


class MessageSender(eventing.EventTarget):

    def __init__(self, addr, signal, atlas):
        assert addr and signal and  atlas
        self.addr = addr
        self.signal = signal
        self.atlas = atlas

    @classmethod
    def factory(cls):
        addr = store.getState('addr')
        signal = hub.SignalClient.factory()
        atlas = hub.AtlasClient.factory()
        return cls(addr, signal, atlas)

    async def makeAttachmentPointer(self, attachment):
        assert isinstance(attachment, Attachment)
        key = secrets.token_bytes(64)
        ptr = protobufs.AttachmentPointer()
        ptr.key = key
        ptr.contentType = attachment.type
        iv = secrets.token_bytes(16);
        encrypted = crypto.encryptAttachment(attachment.serialize(), key, iv)
        ptr.id = await self.signal.putAttachment(encrypted)
        return ptr

    async def uploadAttachments(self, message):
        if message.attachments:
            uploads = map(message.attachments, self.makeAttachmentPointer)
            message.attachmentPointers = await asyncio.gather(uploads)

    async def send(self,
        to=None, distribution=None,
        text=None, html=None, body=None,
        data=None,
        threadId=None,
        threadType='conversation',
        threadTitle=None,
        messageType='content',
        messageId=None,
        messageRef=None,
        expiration=None,
        attachments=None,
        flags=None,
        sendTime=datetime.datetime.now(),
        userAgent='librelay-python'):
        if data is None:
            data = {}
        if threadId is None:
            threadId = str(uuid.uuid4())
        if messageId is None:
            messageId = str(uuid.uuid4())
        if distribution is None:
            if to is None:
                raise TypeError("`to` or `distribution` required")
            distribution = await self.atlas.resolveTags(to)
        if body is None:
            body = []
        if text:
            body.append({
                "type": 'text/plain',
                "value": text
            })
        if html:
            body.append({
                "type": 'text/html',
                "value": html
            })
        if body:
            data['body'] = body
        if attachments:
            data['attachments'] = [x.getMeta() for x in attachments]
        timestamp = msnow()
        msg = Message(
            addrs=distribution['userids'],
            threadId=threadId,
            body=[{
                "version": 1,
                "threadId": threadId,
                "threadType": threadType,
                "messageId": messageId,
                "messageType": messageType,
                "messageRef": messageRef,
                "distribution": {
                    "expression": distribution['universal']
                },
                "sender": {
                    "userId": self.addr
                },
                "sendTime": sendTime.isoformat(),
                "userAgent": userAgent,
                "data": data
            }],
            timestamp = timestamp,
            attachments = attachments,
            expiration = expiration,
            flags = flags)
        await self.uploadAttachments(msg)
        msgProto = msg.toProto()
        await self._sendSync(msgProto, timestamp, threadId,
                             expiration and timestamp)
        return await self._send(msgProto, timestamp,
                                self.scrubSelf(distribution['userids']))

    async def _send(self, msgProto, timestamp, addrs):
        assert all(addrs)
        logger.debug(f"Sending to: {addrs}")
        m = OutgoingMessage(self.signal, timestamp, msgProto)
        m.on('keychange', self.onKeyChange)
        for res in asyncio.as_completed([m.sendToAddr(x) for x in addrs]):
            try:
                await res
            except Exception as e:
                logger.exception('Message send error')
                await self.onError(e)
        return m

    async def onError(self, e):
        ev = eventing.Event('error')
        ev.error = e
        await self.dispatchEvent(ev)

    async def onKeyChange(self, e):
        await self.dispatchEvent(eventing.KeyChangeEvent(e))

    async def _sendSync(self, content, timestamp, threadId,
                        expirationStartTimestamp):
        content = protobufs.Content()
        sent = content.syncMessage.sent
        sent.timestamp = timestamp
        sent.message.CopyFrom(content.dataMessage)
        if threadId:
            sent.destination = threadId
        if expirationStartTimestamp:
            sent.expirationStartTimestamp = expirationStartTimestamp
        return await self._send(content, timestamp, [self.addr])

    async def syncReadMessages(self, reads):
        if not reads:
            logger.warning("No reads to sync")
        readProtobufs = []
        for x in reads:
            buf = protobufs.SyncMessage.Read()
            buf.timestamp = x.timestamp
            buf.sender = x.sender
            readProtobufs.append(buf)
        syncMessage = protobufs.SyncMessage()
        syncMessage.read = readProtobufs
        content = protobufs.Content()
        content.syncMessage = syncMessage
        return await self._send(content, msnow(), [self.addr])

    def scrubSelf(self, addrs):
        return [x for x in addrs if x != self.addr]

    async def closeSession(self, addr, timestamp=None):
        if timestamp is None:
            timestamp = msnow()
        dataMessage = protobufs.DataMessage()
        dataMessage.flags = protobufs.DataMessage.END_SESSION
        content = protobufs.Content()
        content.dataMessage = dataMessage
        outmsg = await self._send(content, timestamp, [addr])
        deviceIds = store.getDeviceIds(addr)
        raise NotImplementedError('xxx')
        #await Promise(resolve => {
        #    outmsg.on('sent', resolve)
        #    outmsg.on('error', resolve)
        #})
        #await Promise.all(deviceIds.map(deviceId => {
        #    address = libsignal.SignalProtocolAddress(addr, deviceId)
        #    sessionCipher = libsignal.SessionCipher(store, address)
        #    return sessionCipher.closeOpenSessionForDevice()
        #}))
